### PR TITLE
Port towards config file-based training

### PR DESCRIPTION
**Pull request by naga-karthik**
This PR heavily improves the way we train the contrast-agnostic model. Earlier, a lot of important arguments were passed through the command line, making the arguments list ever increasing and messy. With this PR, all the important params (and hyperparams) are specified in `.yaml` file that could be used for training. This could help other segmentation projects looking to use monai-based models to improve upon or replace nnUNet model. 

Usage is now as simple as:
```
python monai/main.py --model nnunet --config configs/train_soft_all.yaml
```

This improved training strategy is mainly motivated by #87 - a follow-up project looking at the model drift (using lifelong learning) when training the contrast-agnostic model by adding more datasets. This also makes the training easier for a foundational model for SC segmentation that we were discussing with @valosekj .

### CODE REVIEW

File: scripts/train.sh [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  --contrast $CONTRAST \
+  --label-type $LABEL_TYPE \
+  --seed 15
+
+echo "----------------------------------------"
+echo "Training contrast-agnostic '$MODEL' model on '$CONTRAST' contrasts with '$LABEL_TYPE' labels"
+echo "----------------------------------------"
+
+# Train the model
+python monai/main.py --model $MODEL --config $PATH_TRAIN_YAML --debug
```
Comment (valosekj) [inline scripts/train.sh:71]: Not sure if this is necessary or not, but I usually use `conda deactivate` as the last command.

File: scripts/train.sh [hunk 2]
```diff
@@ -0,0 +1,71 @@
+#!/bin/bash
+# 
+# This script does the following:
+# 1. Creates a virtual environment and installs the required dependencies
```
Comment (valosekj) [inline scripts/train.sh:4]: Maybe you can specify that the env must be named `venv_monai`

File: csa_qc_evaluation_spine_generic/comparison_across_thresholds.sh [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+# Usage:
+#     sct_run_batch -config config.json
+#
+# Example of config.json:
+# {
+#  "path_data"   : "<PATH_TO_DATASET>",
+#  "path_output" : "<PATH_TO_DATASET>_2023-08-18",
+#  "script"      : "<PATH_TO_REPO>/model_seg_sci/baselines/comparison_with_other_methods.sh",
+#  "jobs"        : 8,
+#  "script_args" : "<PATH_TO_REPO>/model_seg_sci/packaging/run_inference_single_subject.py <PATH_TO_MODEL>/sci-multisite-model <PATH_TO_CONTRAST-AGNOSTIC_REPO>/monai/run_inference_single_image.py <PATH_TO_CONTRAST-AGNOSTIC_MODEL>"
```
Comment (valosekj) [inline csa_qc_evaluation_spine_generic/comparison_across_thresholds.sh:17]: Very minor nitpick: you could update `model_seg_sci`

File: csa_qc_evaluation_spine_generic/comparison_across_training_labels.sh [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+# Usage:
+#     sct_run_batch -config config.json
+#
+# Example of config.json:
+# {
+#  "path_data"   : "<PATH_TO_DATASET>",
+#  "path_output" : "<PATH_TO_DATASET>_2023-08-18",
+#  "script"      : "<PATH_TO_REPO>/model_seg_sci/baselines/comparison_with_other_methods.sh",
+#  "jobs"        : 8,
+#  "script_args" : "<PATH_TO_REPO>/model_seg_sci/packaging/run_inference_single_subject.py <PATH_TO_MODEL>/sci-multisite-model <PATH_TO_CONTRAST-AGNOSTIC_REPO>/monai/run_inference_single_image.py <PATH_TO_CONTRAST-AGNOSTIC_MODEL>"
```
Comment (valosekj) [inline csa_qc_evaluation_spine_generic/comparison_across_training_labels.sh:16]: Very minor nitpick: you could update `model_seg_sci`

File: processing_spine_generic/generate_soft_bin_labels.sh [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+# Usage:
+#     sct_run_batch -config config.json
+#
+# Example of config.json:
+# {
+#  "path_data"   : "<PATH_TO_DATASET>",
+#  "path_output" : "<PATH_TO_DATASET>_2023-08-18",
+#  "script"      : "<PATH_TO_REPO>/model_seg_sci/baselines/comparison_with_other_methods.sh",
+#  "jobs"        : 8,
+#  "script_args" : "<PATH_TO_REPO>/model_seg_sci/packaging/run_inference_single_subject.py <PATH_TO_MODEL>/sci-multisite-model <PATH_TO_CONTRAST-AGNOSTIC_REPO>/monai/run_inference_single_image.py <PATH_TO_CONTRAST-AGNOSTIC_MODEL>"
```
Comment (valosekj) [inline processing_spine_generic/generate_soft_bin_labels.sh:16]: Very minor nitpick: you could update `model_seg_sci`

File: qc_other_datasets/generate_qc.sh [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+# Usage:
+#     sct_run_batch -config config.json
+#
+# Example of config.json:
+# {
+#  "path_data"   : "<PATH_TO_DATASET>",
+#  "path_output" : "<PATH_TO_DATASET>_2023-08-18",
+#  "script"      : "<PATH_TO_REPO>/model_seg_sci/baselines/comparison_with_other_methods.sh",
+#  "jobs"        : 8,
+#  "script_args" : "<PATH_TO_REPO>/model_seg_sci/packaging/run_inference_single_subject.py <PATH_TO_MODEL>/sci-multisite-model <PATH_TO_CONTRAST-AGNOSTIC_REPO>/monai/run_inference_single_image.py <PATH_TO_CONTRAST-AGNOSTIC_MODEL>"
```
Comment (valosekj) [inline qc_other_datasets/generate_qc.sh:16]: Very minor nitpick: you could update `model_seg_sci`

File: qc_other_datasets/generate_qc_epi.sh [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+# Usage:
+#     sct_run_batch -config config.json
+#
+# Example of config.json:
+# {
+#  "path_data"   : "<PATH_TO_DATASET>",
+#  "path_output" : "<PATH_TO_DATASET>_2023-08-18",
+#  "script"      : "<PATH_TO_REPO>/model_seg_sci/baselines/comparison_with_other_methods.sh",
+#  "jobs"        : 8,
+#  "script_args" : "<PATH_TO_REPO>/model_seg_sci/packaging/run_inference_single_subject.py <PATH_TO_MODEL>/sci-multisite-model <PATH_TO_CONTRAST-AGNOSTIC_REPO>/monai/run_inference_single_image.py <PATH_TO_CONTRAST-AGNOSTIC_MODEL>"
```
Comment (valosekj) [inline qc_other_datasets/generate_qc_epi.sh:16]: Very minor nitpick: you could update `model_seg_sci`

File: scripts/train.sh (hunk 2 above)
Comment (naga-karthik) [reply scripts/train.sh]: Done! https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/pull/90/commits/d37f25524701052bab7fe94739b5e1fa63e908ef

File: processing_spine_generic/generate_soft_bin_labels.sh (hunk 5 above)
Comment (naga-karthik) [reply processing_spine_generic/generate_soft_bin_labels.sh:16]: done in https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/pull/90/commits/041ba27636d4dfc14b03f37a41bbebc2d6c80627

File: qc_other_datasets/generate_qc_epi.sh (hunk 7 above)
Comment (naga-karthik) [reply qc_other_datasets/generate_qc_epi.sh:16]: done in https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/pull/90/commits/041ba27636d4dfc14b03f37a41bbebc2d6c80627

File: qc_other_datasets/generate_qc.sh (hunk 6 above)
Comment (naga-karthik) [reply qc_other_datasets/generate_qc.sh]: done in https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/pull/90/commits/894db9cf92ec43d11cc2b09a860badc5e4c6e802

File: csa_qc_evaluation_spine_generic/comparison_across_training_labels.sh (hunk 4 above)
Comment (naga-karthik) [reply csa_qc_evaluation_spine_generic/comparison_across_training_labels.sh]: done in https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/pull/90/commits/b728ee7390c80659f52e85d15228e4c7e9f3727e

File: configs/train_soft_all.yaml [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+  results_dir: /home/GRAMES.POLYMTL.CA/u114716/contrast-agnostic/results
+  # Path to the saved wandb logs directory
+  # if None, starts training from scratch. Otherwise, resumes training from the specified wandb run folder
+  wandb_run_folder: None  
+
+dataset:
+  # Dataset name (will be used as "group_name" for wandb logging)
+  name: spine-generic
+  # Path to the dataset directory containing all datalists (.json files)
+  root_dir: /home/GRAMES.POLYMTL.CA/u114716/contrast-agnostic/datalists/spine-generic/seed15
```
[the remark this answers is not in the document]
Comment (naga-karthik) [reply configs/train_soft_all.yaml]: for now, this is out of the scope of the PR, so resolving it. It has been noted as a todo in https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/issues/101#issuecomment-1983746449

**Review by sandrinebedard — APPROVED** [review]

**Pull request merged**